File: dotnet/path_traversal/StaticFiles.cs
using System.IO;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("[controller]")]
public class StaticFilesController : ControllerBase
{
    private const string StaticDir = "/var/www/static";

    // VULN 1: Directory.GetFiles with user-controlled directory - exposes arbitrary directory contents
    [HttpGet("browse")]
    public IActionResult BrowseDirectory(string dir = "")
    {
        string target = Path.Combine(StaticDir, dir);
        string[] entries = Directory.GetFiles(target);
        return Ok(entries);
    }

    // VULN 2: File.ReadAllBytes with string interpolation - user controls name and type
    [HttpGet("asset")]
    public IActionResult ServeAsset(string name, string type)
    {
        string filePath = $"{StaticDir}/{type}/{name}";
        byte[] content = System.IO.File.ReadAllBytes(filePath);
        return File(content, "application/octet-stream");
    }

    // VULN 3: File.Copy with user-controlled destination - write to arbitrary path
    [HttpPost("copy")]
    public IActionResult CopyTemplate(string template, string dest)
    {
        string src = Path.Combine(StaticDir, "templates", template);
        System.IO.File.Copy(src, dest, overwrite: true);
        return Ok("copied");
    }
}

File: java/path_traversal/StaticFiles.java
import jakarta.servlet.http.*;
import java.io.*;
import java.nio.file.*;
import java.util.stream.Collectors;

public class StaticFiles extends HttpServlet {
    private static final String STATIC_DIR = "/var/www/static";

    // VULN 1: Files.list with user-controlled directory - exposes arbitrary directory contents
    protected void doBrowse(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String subdir = req.getParameter("dir") != null ? req.getParameter("dir") : "";
        Path target = Paths.get(STATIC_DIR).resolve(subdir);
        String entries = Files.list(target)
            .map(p -> p.getFileName().toString())
            .collect(Collectors.joining(","));
        resp.getWriter().write(entries);
    }

    // VULN 2: Files.readAllBytes with string.format path - user controls name and type
    protected void doServeAsset(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String name = req.getParameter("name");
        String type = req.getParameter("type");
        String filePath = String.format("%s/%s/%s", STATIC_DIR, type, name);
        byte[] content = Files.readAllBytes(Paths.get(filePath));
        resp.getOutputStream().write(content);
    }

    // VULN 3: Files.copy with user-controlled destination - write to arbitrary path
    protected void doCopyTemplate(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String template = req.getParameter("template");
        String dest = req.getParameter("dest");
        Path src = Paths.get(STATIC_DIR, "templates", template);
        Files.copy(src, Paths.get(dest), StandardCopyOption.REPLACE_EXISTING);
        resp.getWriter().write("copied");
    }
}
